Userland/Applications/PixelPaint/PaletteWidget.cpp

```diff
@@ -162,4 +136,87 @@ void PaletteWidget::set_secondary_color(Color color)
     m_secondary_color_widget->update();
 }
 
+Result<void, String> PaletteWidget::load_palette_file(String const& file_path)
+{
+    auto file_or_error = Core::File::open(file_path, Core::OpenMode::ReadOnly);
+    if (file_or_error.is_error())
+        return file_or_error.error();
+
+    auto& file = *file_or_error.value();
+
+    Vector<Color> new_colors;
+
+    while (file.can_read_line()) {
+        auto line = file.read_line();
+        auto color = Color::from_string(line);
+        if (!color.has_value()) {
+            return String::formatted("Could not parse \"{}\" as a color", line);
+        }
+
+        new_colors.append(color.value());
+    }
+
+    file.close();
+
+    populate_color_container(new_colors);
+
+    return {};
+}
+
+Result<void, String> PaletteWidget::save_palette_file(String const& file_path)
+{
+    auto file_or_error = Core::File::open(file_path, Core::OpenMode::WriteOnly);
+    if (file_or_error.is_error())
+        return file_or_error.error();
+
+    auto& file = *file_or_error.value();
+
+    for (auto& color_container : m_color_container->child_widgets()) {
+        color_container.for_each_child_of_type<ColorWidget>([&](auto& color_widget) {
+            file.write(color_widget.color().to_string_without_alpha());
+            file.write("\n");
+            return IterationDecision::Continue;
+        });
+    }
+
+    file.close();
+
+    return {};
+}
+
+void PaletteWidget::populate_color_container(Vector<Color> const& colors)
+{
+    int colors_to_add = colors.size();
+    if (colors_to_add == 0) {
+        dbgln("No colors to populate the container with :(");
+        return;
+    }
```

Comment:
Is this really an error or do we want to allow empty palettes ?

Reply:
Not sure if we would want to allow that, since there is not a way to add color dials to the palette widget via GUI (yet). So loading a palette file without colors would potentially lead to an empty palette widget. I thought it wise to not replace the existing palette in that case.

Reply:
Okay. In that case, you probably want a fallback in the PaletteWidget constructor when `result.is_error()` is true that adds e.g a black and a white color. (right now, if there's no palette, it will not load anything but keep running after the popup).